bitcoin/src/merkle_tree/mod.rs

```diff
@@ -39,12 +39,29 @@ hashes::hash_newtype! {
 impl_hashencode!(TxMerkleNode);
 impl_hashencode!(WitnessMerkleNode);
 
-impl From<Txid> for TxMerkleNode {
-    fn from(txid: Txid) -> Self { Self::from_byte_array(txid.to_byte_array()) }
+/// A trait that defines a hash that can appear as a node in a Merkle tree.
+pub trait MerkleNode: Hash + Encodable {
```

Comment:
In a718079ecbf32359cbd2b4412adcccd2d533d1e1:

This trait should be inverted. It should be implemented on the node types, have `type Leaf` (which I don't think needs any bounds) and a `from_leaf` method.

Right now it is confusingly named because it's called `MerkleNode` but it's implemented on the *leaves* of the merkle tree, not the nodes.

Reply:
Yeah, after sleeping on this I'm not happy with it. The reasons you state are valid, I tried the other way and it also has problems. Its problematic because of the recursive function coupled with the inline function, because of inline we can't change the type (without unsafe) and because of the recursive function the leaf and internal nodes need to be the same type.

Second thing I woke up thinking, this is not going to work with the make-txid-not-a-general-purpose-hash thing, it might using `to/from_byte_array` but I'm not sure. I'm starting to think that we should just have two separate functions with suffix `_txid` and `_wtxid` and just be done with it. Like you say, this merkle construct is unique to merkelising transaction ids.

Reply:
> Its problematic because of the recursive function coupled with the inline function, because of inline we can't change the type (without unsafe) and because of the recursive function the leaf and internal nodes need to be the same type.

Possibly you need a recursive function combined with a thin wrapper function.

> Second thing I woke up thinking, this is not going to work with the make-txid-not-a-general-purpose-hash thing, it might using `to/from_byte_array` but I'm not sure

It will. `to/from_byte_array` is how you "cast" hashes between each other, which is exactly what's happening here. It looks ugly because it's a bad idea that resulted in these Merkle trees having bad behavior on-chain. So that is a win for our new API to make the ugly thing look ugly.

Reply:
I can take this one on if you're tired of it.

Reply:
I did it in https://github.com/rust-bitcoin/rust-bitcoin/pull/2864. I didn't have any issue with recursion and definitely did not need unsafe code.

Reply:
Oh sweet, thanks man.